### PR TITLE
Use `std::optional<>` rather than `boost::optional<>`

### PR DESCRIPTION
**Pull request by christianparpart**
refs #7259

### CODE REVIEW

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `9ee1e5402a89b37cf6b02c7855f23bdf74c03b7c`:
```
Error: Trailing whitespace found:
 libsolidity/parsing/Parser.cpp:932: auto stateMutability = elemTypeName.token() == Token::Address 

```
Please check that your changes are working as intended.

**Comment by erak** [comment]
`b_osx` should be fixed with https://github.com/ethereum/solidity/pull/7554, which contains https://github.com/ethereum/solidity/pull/7552.

**Comment by erak** [comment]
Ah, and did have a chance to look into the Emscripten build?

File: libsolidity/parsing/Parser.cpp [hunk 1]
```diff
@@ -1651,7 +1653,7 @@ ASTPointer<Expression> Parser::parseLeftHandSideExpression(
 		nodeFactory.markEndPosition();
 		auto expressionType = nodeFactory.createNode<ElementaryTypeName>(
 			ElementaryTypeNameToken(Token::Address, 160, 0),
-			boost::make_optional(StateMutability::Payable)
+			std::optional{StateMutability::Payable}
```
Comment (erak) [inline libsolidity/parsing/Parser.cpp]: It looks like the Emscripten build fails here.

File: libsolidity/analysis/ReferencesResolver.cpp [hunk 2]
```diff
@@ -281,6 +281,7 @@ bool ReferencesResolver::visit(InlineAssembly const& _inlineAssembly)
 	yul::ExternalIdentifierAccess::Resolver resolver =
 	[&](yul::Identifier const& _identifier, yul::IdentifierContext, bool _crossesFunctionBoundary) {
 		auto declarations = m_resolver.nameFromCurrentScope(_identifier.name.str());
+		//printf("identifier: \"%s\"\n", _identifier.name.str().c_str());
```
Comment (erak) [inline libsolidity/analysis/ReferencesResolver.cpp]: Leftover.

**Review by erak — CHANGES_REQUESTED** [review]
Left a tiny comment. And we need to figure out, why Emscripten fails.

File: libsolidity/parsing/Parser.cpp (hunk 1 above)
Comment (christianparpart) [reply libsolidity/parsing/Parser.cpp]: The Emscripten's c++ library does not seem to provide the deduction guide for std::optional, that's however [part of C++17](https://en.cppreference.com/w/cpp/utility/optional/deduction_guides).

**Comment by chriseth** [comment]
Why not target the develop branch?

**Comment by christianparpart** [comment]
> Why not target the develop branch?

Sure, I could do that *BUT* I assumed that due to changing 3rd party library code, it *may* potentially be a little more risky, and using a new major release gives enough head-room. Should I rebase on top of `develop`? (I also assumed we're about to release 0.6 rather soon) :)

**Comment by erak** [comment]
I'd also feel more confident with targeting 0.6.0, which should us give some more time _if_ there any unforseen issues as well.

**Comment by christianparpart** [comment]
> Why not target the develop branch?

@chriseth rebased on `develop` branch now.

**Review by chriseth — APPROVED** [review]

**Pull request merged**